### PR TITLE
Refactor LowerArrayToStream

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

**Comment by tpoterba** [comment]
This may not be entirely correct, let's think about it tomorrow. But I think the structure here makes it a bit easier to follow.

**Comment by akotlar** [comment]
sounds good

File: hail/src/main/scala/is/hail/expr/LowerArrayToStream.scala [hunk 1]
```diff
@@ -17,85 +17,46 @@ import is.hail.expr.types.virtual.{TArray, TStream, TStreamable}
...
-      case ArrayFor(a, n, b) => ArrayFor(streamify(a), n, b)
-      case ArrayFold2(a, accum, valueName, seq, result) => ArrayFold2(streamify(a), accum, valueName, seq, result)
-      case x: ApplyIR => apply(x.explicitNode)
-      case RunAggScan(a, name, init, seq, res, sig) => RunAggScan(streamify(a), name, init, seq, res, sig)
-      case _ => unstreamify(Copy(node, Children(node).map { case c: IR => apply(c) }))
-    }
-
-    println(s"ending LowerArrayToStream with: ${ r }\n\n")
-    r
+    streamify(node)
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/LowerArrayToStream.scala:60]: I like this solution. Previously, relying on Copy and Children, we needed to split some nodes out, like ArrayFold, ArrayFold2 to streamify too many arguments. The other issue was knowing when to unstreamify something. That behavior was hard to follow (sometimes we streamify an IR argument, sometimes we apply a child, and we have ToStream, ToArray in both streamify and unstreamiffy. I wasn't successful however having ToStream and ToArray in just one of the two; needed behavior was different. I'll try this version out first thing in the morning.

**Review by akotlar — APPROVED** [review]
approving to iterate one my branch
